qiskit/algorithms/amplitude_estimators/ae.py

```diff
@@ -386,13 +386,13 @@ class AmplitudeEstimationResult(AmplitudeEstimatorResult):
 
     def __init__(self) -> None:
         super().__init__()
-        self._num_evaluation_qubits = None
-        self._mle = None
-        self._mle_processed = None
-        self._samples = None
-        self._samples_processed = None
-        self._y_measurements = None
-        self._max_probability = None
+        self._num_evaluation_qubits: Optional[int] = None
+        self._mle: Optional[float] = None
+        self._mle_processed: Optional[float] = None
+        self._samples: Optional[Dict[float, float]] = None
```

Comment:
This seems to be a common change here. Both algorithms and opflow used to run against mypy in Aqua before they were migrated to Terra here. Since then mypy has not been run here and some code has been added so I would have expected some issues. But mypy always internally figured the type of instance variables based on its usage so unless things were more complicated around the types, where it locked in one aspect and it was used with a different type later, in general this was not needed. @manoelmarques since you dealt more directly with mypy in this area any comment.

Reply:
I went with error messages, somehow mypy couldn't figure out that it's `Optional[int]` rather than `NoneType`, and thus assignment of `int` to this variable is leading to type error

Reply:
What version of mypy are you running with? I know mypy has put out new versions since we ran the code when it was in Aqua. Here for instance is a [mypy ini](https://github.com/Qiskit/qiskit-nature/blob/main/mypy.ini) file, for code that was migrated from Aqua, where in the applications repos we run mypy, and is updated from what we used to do in Aqua. For instance we have [strict_optional ](https://mypy.readthedocs.io/en/stable/config_file.html#confval-strict_optional) set False, which allows None to be any type. What settings are you running this with. I assume you are doing this all locally since I do not see requirements-dev.txt nor CI updated to run this here yet. Maybe its too soon.

Reply:
I'm running `mypy 0.970` with following command `dmypy run -- --ignore-missing-imports --no-strict-optional --allow-untyped-globals --allow-redefinition --show-error-codes qiskit` taken from https://github.com/Qiskit/qiskit-terra/issues/6905#issuecomment-1130009776
I'm indeed doing it locally, since there are still ~1000 mypy errors after this PR (and ~1400 before). Indeed fixing everything in one PR is too much, and even this PR will probably be broken into a couple of smaller ones, as discussed a bit earlier here.